### PR TITLE
Fix edge-case contiguity mismatch for Allgatherv

### PR DESCRIPTION
**Pull request by ClaudiaComito**
## Description

Function `communication.mpi_type_and_elements_of()` calculates type and number of elements that will be sent/received in an `Allgatherv` call. How to calculate the number of elements depends on whether the input object (most likely a torch Tensor) is contiguous.  

In some edge cases,  i.e. in case of singleton split dimension, `torch.Tensor.is_contiguous()` might return `True` on a process while it is `False` on others. This result in a mismatch of the send/recv elements among processes in Allgatherv and resulting deadlock (see #1057 ).

Issue/s resolved: #1057 

## Changes proposed:
- Introduce global `is_contiguous` boolean as kwarg for `communication.as_buffer()` and `communication.mpi_type_and_elements_of()`. It is set to False on all processes if object dimensions have been permuted, independently on the size of the dimension
- More deadlocks/errors came up running the tests, solved here:
  - removed unnecessary MPI calls for non-distributed cases in `dndarray.resplit_()`, `manipulations.resplit()`, `linalg.matmul()`,  and tests
  - in `test_suites.basic_tests.TestCase.assert_array_equal`, local tensors are compared to the relevant slices of the numpy reference array, instead of gathering the distributed DNDarray every time. TODO: Same should be implemented in #1064 
  - `ht.allclose` now works on operands with different dtype as well. Related to  #889 

## Type of change

- Bug fix (non-breaking change which fixes an issue)
## Memory requirements

NA
## Due Diligence
- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Title of PR is suitable for corresponding CHANGELOG entry

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

**Review by mtar — COMMENTED** [review]
Thank you @ClaudiaComito 

You added extra lines for arrays with different splits when the number of MPI processes is one. It doesn't make much sense to pass an argument when the split won't take place on one process. What do you think about disallowing splits or automatically setting he value to `None` when only a single process is involved at array creation time? It would save us some tests/checks.

File: heat/core/communication.py [hunk 1]
```diff
@@ -255,12 +259,18 @@ def mpi_type_and_elements_of(
...
         # ToDo: The option to explicitely specify the counts and displacements to be send still needs propper implementation
         """
         mpi_type, elements = cls.__mpi_type_mappings[obj.dtype], torch.numel(obj)
 
-        # simple case, continuous memory can be transmitted as is
-        if obj.is_contiguous():
+        # simple case, contiguous memory can be transmitted as is
+        if is_contiguous is None:
+            # determine local contiguity
+            is_contiguous = obj.is_contiguous()
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/communication.py:271]: Thanks @mtar,  that's a great question. The obvious case in which this might happen is the permutation, and this is dealt with in this PR. Outside of that, we're simply falling back to the previous implementation. 

I could add a global check that sets `is_contiguous` to False if the local contiguities are dishomogeneous.

[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/communication.py:271]: > Thank you @ClaudiaComito
> 
> You added extra lines for arrays with different splits when the number of MPI processes is one. It doesn't make much sense to pass an argument when the split won't take place on one process. What do you think about disallowing splits or automatically setting he value to `None` when only a single process is involved at array creation time? It would save us some tests/checks.

This is a general discussion worth having, maybe not re: this bug fix. 
My main argument against setting all splits to None when running on 1 MPI process, is that it will be confusing for users while they are testing their code (potentially on 1 process or even interactively). 

Anyway, let's discuss it in a separate Issue. 

As far as I'm concerned, I'm done with this PR.

[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/communication.py:271]: > Thanks @mtar, that's a great question. The obvious case in which this might happen is the permutation, and this is dealt with in this PR. Outside of that, we're simply falling back to the previous implementation.
> 
> I could add a global check that sets `is_contiguous` to False if the local contiguities are dishomogeneous.

I've decided not to add (yet another) global check for contiguous status for now, as I can't think of the appropriate edge-case to test it. We are already testing for column-first memory layout operations. If anybody can think of something, let me know.

**Review by mtar — APPROVED** [review]
